### PR TITLE
LCL history request and response.

### PR DESCRIPTION
**Pull request by asanka-indrajith**
Detect and request missing lcl history from another random node.
Sending lcl history response to a asked node.
Getting lcl history response and applying it.
Delete lcl that exceeds max ledger sequence.

### CODE REVIEW

File: src/cons/cons.cpp [hunk 1]
```diff
@@ -123,6 +110,21 @@ void consensus()
     }
     else // Stage 1, 2, 3
     {
+
+        std::cout << "Started stage " << std::to_string(ctx.stage) << "\n";
```
Comment (ravidulashan) [inline src/cons/cons.cpp]: LOG_DBG

File: src/cons/cons.cpp [hunk 2]
```diff
@@ -533,8 +536,9 @@ void timewait_stage(const bool reset)
  */
 void apply_ledger(const p2p::proposal &cons_prop)
 {
-    ctx.led_seq_no++;
-    ctx.lcl = cons::save_ledger(cons_prop, ctx.led_seq_no);
+    const std::tuple<const uint64_t, std::string> new_lcl = save_ledger(cons_prop);
```
Comment (ravidulashan) [inline src/cons/cons.cpp]: since you are assigning two data types you can use std::pair which is optimised to hold two data types

Comment (asanka-indrajith) [reply src/cons/cons.cpp]: I don't think there's any additional benefits using pair over tuple. In fact tuple gives more flexibility.

File: src/cons/ledger_handler.cpp [hunk 3]
```diff
@@ -4,60 +4,139 @@
...
+    for (itr = cons::ctx.lcl_list.begin();
+         itr != cons::ctx.lcl_list.lower_bound(led_seq_no);
+         itr++)
+    {
+        const std::string file_name = itr->second;
+        std::string file_path;
+        file_path.reserve(dir_path.size() + itr->second.size() + 4);
+        file_path.append(dir_path);
+        file_path.append(file_name);
+        file_path.append(".lcl");
```
Comment (ravidulashan) [inline src/cons/ledger_handler.cpp]: file_path.append(dir_path).append(file_name).append(".lcl")

File: src/p2p/p2p.cpp [hunk 4]
```diff
@@ -71,7 +71,8 @@ void peer_connection_watchdog()
             }
         }
 
-        util::sleep(conf::cfg.roundtime * 4);
+        //util::sleep(conf::cfg.roundtime * 4);
+        util::sleep(200);
```
[the remark this answers is not in the document]
Comment (asanka-indrajith) [reply src/p2p/p2p.cpp]: Yes. This is to dev test consensus. We'll need to come up with more suitable configuration(like previous code line) in future when we finalize consensus.

File: src/cons/ledger_handler.hpp [hunk 5]
```diff
@@ -6,15 +6,35 @@
 namespace cons
 {
 
+//max ledger count
+static const uint64_t MAX_LEDGER_SEQUENCE = 250;
```
[the remark this answers is not in the document]
Comment (asanka-indrajith) [reply src/cons/ledger_handler.hpp]: constexpr

**Pull request merged**